machine/samd21: basic implementation for DAC

deadprogram opened the pull request:
This PR adds a very basic DAC implementation for the SAMD21. It does not address any timer/DMA functionality needed to create a complex waveform, that will be for future PRs.

deadprogram committed:
machine/samd21: some WIP on a DAC implementation

Signed-off-by: Ron Evans <ron@hybridgroup.com>

deadprogram committed:
machine/samd21: cleanup of minimalist implementation for DAC

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram committed:
build: add circuitplayground express DAC example to smoke tests

Signed-off-by: deadprogram <ron@hybridgroup.com>

aykevl commented on src/examples/dac/dac.go:
```diff
@@ -0,0 +1,40 @@
+	}
+}
+
+func play(val uint16) {
```
Why 'play'? Just wondering why this particular name, it makes me think of something related to audio, which this isn't.

deadprogram replied:
Actually, this example does play an audio tone at 6 different amplitudes using the Circuit Playground Express built-in speaker.

aykevl replied:
Huh that's strange. What kind of speaker is that? Wouldn't a speaker normally need a sine wave (or something) to play a tone instead of a particular voltage?

deadprogram replied:
The loop turns the power on and off acting like a very simple wave ala PWM. Of course just setting a single voltage would not produce any sound! :)

aykevl commented on src/machine/machine_atsamd21.go:
```diff
@@ -2209,3 +2209,69 @@ func ResetProcessor() {
+
+// Write is intended to satisfy the Writer interface.
+// This current implementation is probably not very useful.
+func (dac DAC) Write(p []uint16) (n int, err error) {
```
This does not satisfy the `Writer` interface, in fact it breaks this support because the name is the same while the signature is different. Maybe it's better to leave it out until a good use case can be found.

deadprogram replied:
Removed.

aykevl commented on src/machine/machine_atsamd21.go:
```diff
@@ -2209,3 +2209,69 @@ func ResetProcessor() {
+}
+
+// Close the DAC.
+func (dac DAC) Close() error {
```
Probably unnecessary when `Write` doesn't exist (and even then, not sure how it's useful).

deadprogram replied:
Removed.

aykevl commented on src/machine/machine_atsamd21.go:
```diff
@@ -2209,3 +2209,69 @@ func ResetProcessor() {
+
+// WriteData writes a single value to the DAC.
+func (dac DAC) WriteData(value uint16) error {
+	syncDAC()
```
This first `syncDAC` should not be necessary, if the DAC is synchronized after every register access. Maybe a `syncDAC` is missing at the end of the `Configure` method?

deadprogram replied:
You were correct. Removed.

aykevl reviewed (commented):
So, this DAC basically creates a particular voltage (not PWM) on a pin? That's interesting, I haven't seen that often. I wonder if there are other chips with similar functionality, to match the API?

deadprogram committed:
machine/samd21: remove functions related to attempted Writer interface

Signed-off-by: deadprogram <ron@hybridgroup.com>

sago35 commented on src/examples/dac/dac.go:
```diff
@@ -0,0 +1,40 @@
+
+	machine.DAC0.Configure(machine.DACConfig{})
+
+	data := []uint16{4096, 2058, 1024, 512, 256, 128}
```
Since samd21 is a 10-bit DAC, numbers above 1024 are meaningless.
But I want to use the same test with samd51, so I think it is better to leave it as it is.

Samd51 is a 12-bit DAC and its maximum value is 4095.
In addition, since 2058 also looks wrong, isn't 2048 the correct value?

In other words, I think the following is correct.

```go
data := []uint16{4095, 2048, 1024, 512, 256, 128}
```

sago35 commented on src/machine/machine_atsamd21.go:
```diff
@@ -2209,3 +2209,54 @@ func ResetProcessor() {
+func (dac DAC) WriteData(value uint16) error {
+	sam.DAC.DATA.Set(value & 0x3FF)
+	syncDAC()
+	sam.DAC.CTRLA.SetBits(sam.DAC_CTRLA_ENABLE)
```
From reading the datasheet, I don't think the following is necessary
It actually works if you comment out.

```go
	sam.DAC.CTRLA.SetBits(sam.DAC_CTRLA_ENABLE)
	syncDAC()
```

deadprogram replied:
You are quite correct @sago35 now removed. Thank you.

deadprogram committed:
machine/samd21: remove extra DAC enable that is not required

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram commented:
I think all feedback has now been addressed in this PR.

aykevl commented:
The implementation looks good to me. However, I have some concerns about the API.

```go
func (dac DAC) WriteData(value uint16)
```

Similar to the ADC, I would suggest this value should use the full 16-bit scale to be more portable across chips. As stated above, the number of bits varies by chip, but it doesn't seem likely it will be more than 16 bits on common chips. For comparison, the DAC in the stm32f103 is also 12 bits, like the samd51.
I'm assuming the DAC by default uses the VCC (usually 3.3V) as reference voltage for the maximum value?

I'm also not sure about the name, `WriteData` seems a bit generic. What about `Set` for example, analogous to the ADC `Get` function?

deadprogram committed:
machine/samd21: change API for DAC to more closely match other machine APIs.

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram commented:
I have now renamed the function to `Set(value uint16)` and also changed to expect 16-bit values and scale the input value. This does seems like the correct way to handle, @aykevl thanks for the suggestions.

@sago35 you will have to change your PR #1198 a little more than expected once this one has been merged, assuming that the changes are acceptable to everyone.

aykevl reviewed (approved):
LGTM, thanks!

deadprogram commented:
OK commits squashed to keep it clean, and merged. Thanks @aykevl and @sago35 for your reviews. Also @sago35 you are now unblocked to complete #1198
